### ISSUE TITLE
Include ANGLE for win7-* runtimes

### ISSUE THREAD

**Issue by kekekeks** [mono/SkiaSharp#243]
Angle has support even for DX9-based rendering which is preinstalled on Win7. 
![ANGLE](https://i.imgur.com/DZWkJdg.png)

Why isn't it included?

**Comment by mattleibow** [comment]
@kekekeks the simple reason that it wasn't available as a pre-build library. Right nor for UWP I use the MS NuGet. I just haven't gotten round to adding the build steps for ANGLE.

Do you know of a download location for a pre-build library? Or will I be building it from source?

**Comment by kekekeks** [comment]
https://www.dropbox.com/s/ayi8ill098jxh1d/angle-d3d9.zip?dl=0 
Version might be a bit outdated ([chromium/2550](https://github.com/google/angle/tree/chromium/2550)) but it targets D3D9.

**Comment by mattleibow** [comment]
This is something that we will be looking at in the future.

**Comment by mattleibow** [comment]
I am actually a bit concerned about this... If I include this, what happens if the user installs another library that uses it?

Right now, UWP does this and I guess I am lucky... _Technically_ the API is consistent, but still - there may be clashes. There really should be a separate NuGet that everyone can reference...

**Comment by kekekeks** [comment]
Please, don't. We are shipping a patched version of ANGLE that is capable of running properly on top of Microsoft Basic Render Driver (the default one doesn't provide a depth buffer).

**Comment by kekekeks** [comment]
Is it OK to close this issue?

**Comment by mattleibow** [comment]
What about UWP? Do you use ANGLE there?

Also, at some point I will need ANGLE. Maybe skia will get a DX backend - I see some work on that now... For example, WPF does not play well with hosting a GL control and the best way is to use a DX view - using ANGLE.

**Comment by mattleibow** [comment]
On a totally separate note... Since the ANGLE context is assembled, any dll could be used - it does not have to be named the same...

But, on another note... If you already have ANGLE with desktop, have you tried skia with an ANGLE context?

**Comment by mattleibow** [comment]
But now I am thinking... if I include ANGLE, it should not conflict.

I will only include it with my WPF and/or WinForms views. So, if you are using WPF, then you are not using Avalonia...

And, if I make a separate NuGet package that is referenced, any potential conflicts can be removed by using ExcludeAssets="all" for the ANGLE package...

What is your bundling process for ANGLE and how does it play with WPF/WinForms?

**Comment by kekekeks** [comment]
Packaging ANGLE as a separate package and referencing it from WPF/Winforms views should work.

**Comment by mattleibow** [comment]
Is this your package: https://www.nuget.org/packages/Avalonia.Angle.Windows.Natives

**Comment by kekekeks** [comment]
Also, consider including [this patch](https://github.com/AvaloniaUI/angle/commit/80a0d41305f513ac933672b4c065db122d573e97), it enables depth buffer for Microsoft Basic Render Device.

**Comment by kekekeks** [comment]
@mattleibow that package doesn't include any patches. Patch is required for [gl-control](https://github.com/AvaloniaUI/Avalonia/pull/3386) branch.

**Comment by kekekeks** [comment]
Also, keep in mind that some graphics drivers (e. g. AMD) ship their own libEGL.dll in system32. So including ANGLE with Skia will break users who are using AMD EGL implementation.

**Comment by mattleibow** [comment]
> Also, consider including this patch, it enables depth buffer for Microsoft Basic Render Device.

You could PR a patch to vcpkg: https://github.com/microsoft/vcpkg/tree/master/ports/angle
Then you can just bump that and not maintain your own fork.

> Also, keep in mind that some graphics drivers (e. g. AMD) ship their own libEGL.dll in system32
Yeah, that might be an issue... How do you handle this? Do you look up in special places first?

**Comment by kekekeks** [comment]
Our current EglDisplay.cs explicitly checks if EGL implementation is provided by ANGLE and skip EGL initialization otherwise, since we don't support running with AMD EGL implementation. But that might be not the case for other users of SkiaSharp.

**Comment by mattleibow** [comment]
I see your discussion on https://groups.google.com/forum/#!topic/angleproject/moZvbwUeAHU

Did you ever submit a patch to ANGLE?

**Comment by kekekeks** [comment]
Nope, gl-control work got kinda frozen since then. Will probably continue soon-ish as we are getting closer to 0.10. The patch should probably explicitly check for the device name, so it won't affect other drivers.

**Comment by msmshazan** [comment]
@kekekeks ANGLE has an extension to use swiftshader backend (https://chromium.googlesource.com/angle/angle/+/76e90947a8e421308dff9d00c4955717c6667f12/extensions/EGL_ANGLE_platform_angle_device_type_swiftshader.txt) wouldn't it be better to use that instead of using 'Microsoft Basic Render Device' when running on VMware ?
